text/0046-metadata-for-offline-signers.md

```diff
@@ -0,0 +1,274 @@
+2. A Merkle tree is constructed from the metadata chunks
+3. A root of tree (as a left element) is merged with Metadata Descriptor (as a right element)
+4. Resulting value is a constant to be included in `additionalSigned` to prove that the metadata seen by cold device is genuine
+
+### Metadata modularization
+
+1. Types registry is stripped from `docs` fields.
+2. Types records are separated into chunks, with enum variants being individual chunks differing by variant index; each chunk consisting of `id` (same as in full metadata registry) and SCALE-encoded 'Type' description (reduced to 1-variant enum for enum variants). Enums with 0 variants are treated as regular types.
+3. Chunks are sorted by `id` in ascending order; chunks with same `id` are sorted by enum variant index in ascending order.
+
+```
+types_registry = metadataV14.types
+modularized_registry = EmptyVector<id, type>
+for (id, type) in types.registry.iterate_enumerate {
+  type.doc = Null
+  if (type is ReduceableEnum) { // false for 0-variant enums
+    for variant in type.variants.iterate {
+      variant_type = Type {
+        path: type.path
+        type_params: Null
+        type_def: TypeDef::Variant(variants: [variant])
+      }
+      modularized_registry.push(id, variant_type)
+    }
+  } else {
+    modularized_registry.push(id, type)
+  }
+}
+
+modularized_registry.sort(|a, b| {
+    if a.id == b.id { //only possible for variants
+      a.variant_index > b.variant_index
+    } else { a.id > b.id }
+  }
+)
+
+```
+
+### Merging protocol
+
+`blake3` transformation of concatenated child nodes (`blake3(left + right)`) as merge procedure;
+
+### Complete Binary Merkle Tree construction protocol
+
+1. Leaves are numbered in ascending order. Leaf index is associated with corresponding chunk.
+2. Merge is performed using the leaf with highest index as right and node with second to highest index as left children; result is pushed to the end of nodes queue and leaves are discarded.
+3. Step (2) is repeated until no leaves or just one leaf remains; in latter case, the last leaf is pushed to the front of the nodes queue.
+4. Right node and then left node is popped from the front of the nodes queue and merged; the result is sent to the end of the queue.
+5. Step (4) is repeated until only one node remains; this is tree root.
+
+```
+Resulting tree for metadata consisting of 5 nodes (numbered from 0 to 4):
+
+       root
+     /     \
+    *       *
+   / \     / \
+  *   0   1   2
+ / \
+3   4
+```
+
+### Digest
+
+1. Blake3 hash is computed for each chunk of modular short metadata registry.
+3. Complete Binary Merkle Tree is constructed as described above.
+4. Root hash of this tree (left) is merged with metadata descriptor blake3 hash (right); this is metadata digest.
+
+Version number and corresponding resulting metadata digest MUST be included into Signed Extensions as specified in Chain Verification section below.
```

Comment:
IMO it should be fine to just say that you use the root hash plus the hash of the metadata descriptor as digest.

Reply:
Are you suggesting removal of line 206? I was kind of afraid this would be a bit difficult to follow if we don't state what's the next step, but yes, we can remove it. Should we?